### PR TITLE
[SPARK-42380][BUILD] Upgrade Apache Maven to 3.9.1

### PR DESCRIPTION
**Pull request by LuciferYang**
### What changes were proposed in this pull request?
This PR aims to upgrade Maven to 3.8.7 from 3.9.1.

These two versions have different http timeout due to the default use of `resolver transport`:
- Maven 3.8.7 use Wagon as default and default timeout is 60000ms
- Maven 3.9.1 use native `transport-http` as default and default timeout is 10000ms

So this pr also add [`-Dmaven.resolver.transport=wagon`](https://maven.apache.org/guides/mini/guide-resolver-transport.html#how-to-upgrade-from-wagon-or-%E2%80%9Cnative-transport-does-not-work%E2%80%9D) user property to `build/mvn` script to stick with Wagon which maven 3.8.7 used.

### Why are the changes needed?
The release notes and as follows:

- https://maven.apache.org/docs/3.9.0/release-notes.html
- https://maven.apache.org/docs/3.9.1/release-notes.html
- https://github.com/apache/maven/releases/tag/maven-3.9.0
- https://github.com/apache/maven/releases/tag/maven-3.9.1

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?

- Pass GitHub Actions
- Manual test :

run `build/mvn -version` wll trigger download `apache-maven-3.9.1-bin.tar.gz`

```
exec: curl --silent --show-error -L https://www.apache.org/dyn/closer.lua/maven/maven-3/3.9.1/binaries/apache-maven-3.9.1-bin.tar.gz?action=download
```

### CODE REVIEW

**Comment by bowenliang123** [comment]
Hi, after bumping to Maven 3.9.1 in Apache Kyuubi (https://github.com/apache/kyuubi/pull/4647), there are notable increasing failures in Maven builds when resolving Maven dependencies from the central mirror.
We are trying to investigate and improve the situation in https://github.com/apache/kyuubi/pull/4692, by increasing Maven Resolver Transport connection timeout back to original default connection timeout 60000ms.

**Comment by LuciferYang** [comment]
The property `aether.connector.connectTimeout` seems nogt working, and there are still maven compilation performance issues, need more investigation ...

**Comment by bowenliang123** [comment]
Agree. Same on kyuubi.

**Comment by dongjoon-hyun** [comment]
Thank you all.

FYI, @steveloughran . Apache Maven 3.9.1 seems to have issues still according to the above discussion.

**Comment by LuciferYang** [comment]
image

The last build took 1 hour and 7 meters, let me re-run GA more times

File: connector/protobuf/pom.xml [hunk 1]
```diff
@@ -92,7 +92,7 @@
         <artifactId>maven-shade-plugin</artifactId>
         <configuration>
           <shadedArtifactAttached>false</shadedArtifactAttached>
-          <shadeTestJar>true</shadeTestJar>
+          <shadeTestJar>false</shadeTestJar>
```
Comment (LuciferYang) [inline connector/protobuf/pom.xml:95]: [22c0173](https://github.com/apache/spark/pull/40738/commits/22c0173793ae5b978103bb2050e66bd72336d86f) hang when maven shade to downloads some non-existent tests. jar and set it to false to try again

File: build/mvn [hunk 2]
```diff
@@ -36,7 +36,7 @@ _DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
 # Preserve the calling directory
 _CALLING_DIR="$(pwd)"
 # Options used during compilation
-_COMPILE_JVM_OPTS="-Xss128m -Xmx4g -XX:ReservedCodeCacheSize=128m"
+_COMPILE_JVM_OPTS="-Xss128m -Xmx4g -XX:ReservedCodeCacheSize=128m -Dmaven.resolver.transport=wagon"
```
Comment (LuciferYang) [inline build/mvn:39]: try to specify `-Dmaven.resolver.transport=wagon`

**Comment by steveloughran** [comment]
noted. interestingly for cutting hadoop releases i have had to turn off connection pooling so as to guarantee reliable builds (docker container having to download **/*.jar from maven central is too brittle otherwise). that's even older versions.

* spark &c with fixed versions should stay on the good ones
* if the latest version of the SBOR is happy on recent maven builds then we can reinstate in hadoop.

File: build/mvn (hunk 2 above)
Comment (bowenliang123) [reply build/mvn:39]: Hi, a kind ping from Kyuubi, we reverted to using Maven 3.8.x in Kyuubi for CI builds and local usage, to reduce side effects and easier to use.

Comment (LuciferYang) [reply build/mvn:39]: Thanks for your sync @bowenliang123

**Comment by srowen** [comment]
I mean, is it worth updating if 3.9.1 introduces small issues that we have to work around? or just wait?

**Comment by LuciferYang** [comment]
Long time no see @srowen , thanks for your review ~

I am not sure if the native http client will change the default timeout value to be same with wagon in the future. But I don't object to letting us wait and see if there will be any changes of the new version of Apache Maven :)

**Comment by LuciferYang** [comment]
@steveloughran Do you have any other questions? I am planning to close this PR and test Apache Maven 3.9.2+ in the future to check if it is possible not to add `-Dmaven.resolver.transport=wagon`. Apache Spark can continue to use 3.8.7 now.

**Comment by steveloughran** [comment]
me, I'm happy with whatever suits, just noting that 3.9.1 was unwelcome trouble for me as it ended up hurting the 3.3.5 release process -not the docker builds, which were still on a lower version, but my local dev builds blew up. joy

**Comment by LuciferYang** [comment]
Thanks for your attention to this pr @bowenliang123 @dongjoon-hyun @srowen @steveloughran, close first ~

**Comment by bowenliang123** [comment]
Thanks for the ping. And welcome to cc me when Spark has the next attempt to use maven 3.9.x/4.x.
I was pessimistic that whether Maven will change the default timeout as it's a default config (`aether.connector.connectTimeout=10000`) of Maven Artifact Resolver not in Maven itself. And as for Maven, it has changed to Artifact Resolver as the default resolver already, which fewer improvements we could expect from that. And we are also difficult to reduce possible timeout to maven central mirrors.

**Comment by dongjoon-hyun** [comment]
Sorry for missing pings, @LuciferYang . I'm back from my vacation.
I'm also +1 for the decision (waiting for new better version). Thank you for closing.

**Pull request closed without merging**